src/libsystemd/sd-device/sd-device.c

```diff
@@ -245,6 +246,28 @@ _public_ int sd_device_new_from_devnum(sd_device **ret, char type, dev_t devnum)
         return sd_device_new_from_syspath(ret, syspath);
 }
 
+_public_ int sd_device_new_from_ifname(sd_device **ret, const char *ifname) {
+        const char *syspath;
+
+        assert_return(ret, -EINVAL);
+        assert_return(ifname, -EINVAL);
+
+        syspath = strjoina("/sys/class/net/", ifname);
+        return sd_device_new_from_syspath(ret, syspath);
+}
+
+_public_ int sd_device_new_from_ifindex(sd_device **ret, int ifindex) {
+        char ifname[IF_NAMESIZE + 1];
+
+        assert_return(ret, -EINVAL);
+        assert_return(ifindex > 0, -EINVAL);
+
+        if (!format_ifname(ifindex, ifname))
+                return -ENODEV;
+
```

Comment:
i think it would be a lot nicer to do this like this:

```
char ifindex_str[1 + DECIMAL_STR_MAX(int) + 1];
xsprintf(ifindex_str, "n%d", ifindex);
return sd_device_new_from_device_id(ret, ifindex_str);
```

That way we don't have to have a separate kernel level translation. This is also nice given that we'll only use /sys/ as source of truth here, instead of possibly rtnetlink or ioctls too. That might be relevant because what you see in /sys/ depends mostly on the mount namespace, but what you see via rtnetlich/ioctls depends on the namespace of the fd you run this on...

Reply:
ah, fuck, i see now that sd_device_new_from_device_id() does exactly this ioctl based lookup internally, so ignore everything i wrote here, as it makes no sense....

kinda sucks that sysfs doesn't expose any interface to lookup interfaces by index. Would make a ton of sense if there was /sys/dev/net/<ifindex> as a symlink pointing to the interface, similar to how there is /sys/dev/block/<major>:</minor> and similar...